src/nomad_simulations/schema_packages/basis_set.py

```diff
@@ -184,22 +184,438 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        | Molden     | (pz, px, py)    | (p₀, p₊₁, p₋₁)              |
+        | CP2K       | (pz, px, py)    | (p₀, p₊₁, p₋₁)              |
+
+        If you select `'Custom'`, specify the explicit order for each ℓ ≥ 1
+        in the optional field `ao_custom_order` below.  Otherwise parsers and
+        writers will assume the canonical order embedded in the named
+        convention.
+        """,
+    )
+
+    ao_custom_order = Quantity(
+        type=JSON,
+        description="""
+        JSON-encoded mapping *ℓ → list-of-labels* that defines a user-specific
+        AO ordering.  Only used when `ao_ordering_convention == "Custom"`.
+
+        Example value::
+
+            {"1": ["pz", "px", "py"],
+            "2": ["d0", "d+1", "d-1", "d+2", "d-2"]}
+
+        Parsers/writers should `json.loads(...)` this string.
+        """,
+    )
+
+    total_number_of_basis_functions = Quantity(
+        type=np.int32,
+        description="""
+        The **total** number of contracted basis functions in this entire set. 
+        This is typically the sum of all `(2l+1)` or cartesian expansions across
+        all shells on all relevant atoms (within the scope of this section).
+        """,
+    )
+
+    functional_compositions = SubSection(
         sub_section=AtomCenteredFunction.m_def, repeats=True
-    )  # TODO change name
+    )
+
+    atomic_orbitals = SubSection(
+        sub_section=AtomicOrbitals.m_def,
+        repeats=False,
+        description='AO layer (cartesian flag, AO→shell map, per-AO normalization).',
+    )
+
+    ecps = SubSection(
+        sub_section=EffectiveCorePotential.m_def,
+        repeats=True,
+        description='Zero or more ECP definitions replacing core electrons.',
+    )
+
+    @check_normalized
+    def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        super().normalize(archive, logger)
+
+        # Set AO count from AO view if present
+        if self.atomic_orbitals is not None and self.atomic_orbitals.num is not None:
+            self.total_number_of_basis_functions = self.atomic_orbitals.num
+
+            if (
+                self.functional_compositions is not None
+                and self.atomic_orbitals.shell_index is not None
+                and len(self.functional_compositions) > 0
+            ):
+                si = self.atomic_orbitals.shell_index
+                try:
+                    max_idx = int(np.max(si)) if len(si) > 0 else -1
+                    if max_idx >= len(self.functional_compositions):
+                        logger.warning(
+                            'shell_index contains index values outside the bounds of functional_compositions.'
+                        )
+                except Exception:
+                    # Log an error for unexpected exception during max_idx check
+                    logger.error(
+                        'Failed to perform max_idx sanity check on shell_index.'
+                    )
```

Comment:
Consider Alvin's `log` decorator. You might have to move this part to its own method if you just want to target that.